Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+                      int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].getCurrentBuffer().digits.emplace_back(MergerDigit{o2::mch::Digit(deId, padId, adc, time),
+                                                                    stopTime, false, solarId, dsAddr, chAddr});
+}
+
+void Merger::mergeDigits(int feeId)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].mergeDigits();
+}
+
+//=======================
+// Data decoder
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
+{
+  size_t ndigits{0};
+
+  uint8_t isStopRDH = 0;
+  uint32_t orbit;
+  uint32_t feeId;
+  uint32_t linkId;
+
+  const auto storeDigit = [&](const Digit& d) {
+    mOutputDigits.emplace_back(d);
+    if (mPrint) {
+      std::cout << "[storeDigit]: digit stored, mOutputDigits size: " << mOutputDigits.size() << std::endl;
+    }
+  };
+
+  auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+    if (mChannelHandler) {
+      mChannelHandler(dsElecId, channel, sc);
+    }
+
+    if (mDs2manu) {
+      channel = ds2manu(int(channel));
+    }
+
+    double digitadc(0);
+    for (auto& s : sc.samples) {
+      digitadc += s;
+    }
+
+    int deId{-1};
+    int dsIddet{-1};
+    if (auto opt = mElec2Det(dsElecId); opt.has_value()) {
+      DsDetId dsDetId = opt.value();
+      dsIddet = dsDetId.dsId();
+      deId = dsDetId.deId();
+    }
+
+    int padId = -1;
+    const Segmentation& segment = segmentation(deId);
+    if ((&segment) == nullptr) {
+      return;
+    }
+
+    padId = segment.findPadByFEE(dsIddet, int(channel));
+    if (mPrint) {
+      auto s = asString(dsElecId);
+      auto ch = fmt::format("{}-CH{:02d}", s, channel);
+      std::cout << ch << "  "
+                << fmt::format("PAD ({:04d} {:04d} {:04d})\tADC {:5.0f}  TIME ({} {} {:02d})  SIZE {}  END {}",
+                               deId, dsIddet, padId, digitadc, orbit, sc.bunchCrossing, sc.sampaTime, sc.nofSamples(), (sc.sampaTime + sc.nofSamples() - 1))
+                << (((sc.sampaTime + sc.nofSamples() - 1) >= 98) ? " *" : "") << std::endl;
+    }
+
+    // skip channels not associated to any pad
+    if (padId < 0) {
+      return;
+    }
+
+    Digit::Time time;
+    time.sampaTime = sc.sampaTime;
+    time.bunchCrossing = sc.bunchCrossing;
+    time.orbit = orbit;
+
+    Digit::Time stopTime;
+    stopTime.sampaTime = sc.sampaTime + sc.nofSamples() - 1;
+    stopTime.bunchCrossing = sc.bunchCrossing;
+    stopTime.orbit = orbit;
+
+    mMerger->addDigit(feeId, static_cast<int>(dsElecId.solarId()), static_cast<int>(dsElecId.elinkId()), static_cast<int>(channel),
+                      deId, padId, digitadc, time, stopTime);
+    ++ndigits;
+  };
+
+  const auto patchPage = [&](gsl::span<const std::byte> rdhBuffer) {
```

Comment:
the patchPage is now doing more that what it says on the tin : it patches the RDH (expected), handle the RDH (assuming there's a rdh handler defined, might be expected, not so sure) and "configure" the merger (unexpected).

Reply:
I agree that the pieces of code that are doing more than patching the RDH contents should be moved outside. I have already created a function for the merger initialization and update, while for the orbit info I propose to wait until #4285 is merged and then port the corresponding code.

Reply:
hum, ok, but then it means #4285 must go first and then this one be rebased on it ?

Reply:
Yes... in any case, I consider #4285 more urgent that this one.